lib/connection.cpp

```diff
@@ -194,37 +195,23 @@ void Connection::resolveServer(const QString& mxidOrDomain)
     }
 
     setHomeserver(maybeBaseUrl);
-    emit resolved();
-    return;
 
-    // FIXME, #178: The below code is incorrect and is no more executed. The
-    // correct server resolution should be done from .well-known/matrix/client
     auto domain = maybeBaseUrl.host();
     qCDebug(MAIN) << "Finding the server" << domain;
-    // Check if the Matrix server has a dedicated service record.
-    auto* dns = new QDnsLookup();
-    dns->setType(QDnsLookup::SRV);
-    dns->setName("_matrix._tcp." + domain);
-
-    connect(dns, &QDnsLookup::finished, [this,dns,maybeBaseUrl]() {
-        QUrl baseUrl { maybeBaseUrl };
-        if (dns->error() == QDnsLookup::NoError &&
-                dns->serviceRecords().isEmpty())
-        {
-            auto record = dns->serviceRecords().front();
-            baseUrl.setHost(record.target());
-            baseUrl.setPort(record.port());
-            qCDebug(MAIN) << "SRV record for" << maybeBaseUrl.host()
-                          << "is" << baseUrl.authority();
+
+    auto job = callApi<GetWellknownJob>();
+    connect(job, &BaseJob::finished, [this, job, maybeBaseUrl] {
+        QUrl baseUrl(job->data().homeserver.baseUrl);
+        if (!baseUrl.isValid() || job->status() != BaseJob::Success) {
+            qCDebug(MAIN) << maybeBaseUrl.host() << "doesn't have valid .well-known file" << "- using the hostname as is";
```

Comment:
This doesn't seem to follow the spec that says to fallback to other means only in case of `BaseJob::NotFound` (404). Admittedly, we don't have a good mechanism for FAIL_PROMPT and FAIL_ERROR at the moment but at the very least getting an invalid URL from the successful job should not fall back to using the serverpart as is.

Reply:
I pushed a commit that tried to follow the spec.